### PR TITLE
Allow propagating pod routes to nodes without overlay network

### PR DESCRIPTION
**Pull request by ScheererJ**
**How to categorize this PR?**

/area networking
/kind enhancement
/kind cleanup

**What this PR does / why we need it**:
Allow propagating pod routes to nodes without overlay network.

Some infrastructures may implement routing in the virtual network by moving all traffic via a routing component. This can introduce additional latency and will eventually become a bottleneck. In case the network is sufficiently capable of using a "short-cut" by applying local routes on the nodes, it can be desirable to do so.

This change allows to trigger propagation of pod routes to all nodes via an option in the networking configuration.

**Which issue(s) this PR fixes**:
None.

**Special notes for your reviewer**:

**Release note**:

```other operator
Allow propagating pod routes to nodes without overlay network by specifying `shoot.spec.networking.providerConfig.overlay.createPodRoutes: true`
```

### CODE REVIEW

File: pkg/controller/actuator_reconcile.go [hunk 1]
```diff
@@ -110,6 +110,13 @@ func (a *actuator) Reconcile(ctx context.Context, _ logr.Logger, network *extens
 			networkConfig.IPv4.Mode = (*calicov1alpha1.IPv4PoolMode)(pointer.String(string(calicov1alpha1.Never)))
 			networkConfig.Backend = (*calicov1alpha1.Backend)(pointer.String(string(calicov1alpha1.None)))
 		}
+		if networkConfig.Overlay != nil && networkConfig.Overlay.CreatePodRoutes != nil {
+			if *networkConfig.Overlay.CreatePodRoutes {
+				networkConfig.Backend = (*calicov1alpha1.Backend)(pointer.String(string(calicov1alpha1.Bird)))
+			} else {
+				networkConfig.Backend = (*calicov1alpha1.Backend)(pointer.String(string(calicov1alpha1.None)))
+			}
+		}
```
Comment (axel7born) [inline pkg/controller/actuator_reconcile.go:119]: Here it is possible to disable `bird` even if `Overlay` is enabled. 
Is that a valid configuration?
The logic in the related Cilium change seems to be different.

Comment (ScheererJ) [reply pkg/controller/actuator_reconcile.go:119]: Yes, indeed. What do you propose?

Comment (DockToFuture) [reply pkg/controller/actuator_reconcile.go:119]: Maybe something like that:

```
if networkConfig.Overlay != nil && !networkConfig.Overlay.Enabled && networkConfig.Overlay.CreatePodRoutes != nil && networkConfig.Overlay.CreatePodRoutes{
networkConfig.Backend = (*calicov1alpha1.Backend)(pointer.String(string(calicov1alpha1.Bird)))
}
```

File: pkg/controller/actuator_reconcile.go [hunk 2]
```diff
@@ -110,6 +110,13 @@ func (a *actuator) Reconcile(ctx context.Context, _ logr.Logger, network *extens
 			networkConfig.IPv4.Mode = (*calicov1alpha1.IPv4PoolMode)(pointer.String(string(calicov1alpha1.Never)))
 			networkConfig.Backend = (*calicov1alpha1.Backend)(pointer.String(string(calicov1alpha1.None)))
```
Comment (DockToFuture) [inline pkg/controller/actuator_reconcile.go:111]: Maybe like that:
```suggestion
if networkConfig.Overlay.CreatePodRoutes != nil && networkConfig.Overlay.CreatePodRoutes{ 
   networkConfig.Backend = (*calicov1alpha1.Backend)(pointer.String(string(calicov1alpha1.Bird))) 
} else {
			networkConfig.Backend = (*calicov1alpha1.Backend)(pointer.String(string(calicov1alpha1.None)))
			}
```

**Comment by DockToFuture** [comment]
Maybe you can also mention the field how this can be configured in the release notes.

File: pkg/controller/actuator_reconcile.go (hunk 1 above)
Comment (ScheererJ) [reply pkg/controller/actuator_reconcile.go:116]: @DockToFuture @axel7born Is this how you expected it?

**Review by DockToFuture — APPROVED** [review]
/lgtm

**Pull request merged**